storehaus-algebra/src/main/scala/com/twitter/storehaus/algebra/MergeableStore.scala

```diff
@@ -57,66 +57,73 @@ object MergeableStore {
     val (collectedMGetResult, getFailures) = collectWithFailures(mGetResult)
 
     val mPutResultFut: Future[Map[K, Future[Option[V]]]] =
-      collectedMGetResult.map { pairs: Seq[(K, (Option[V], Option[V]))] =>
+      collectedMGetResult.map { pairs: Array[(K, (Option[V], Option[V]))] =>
         val pairMap = pairs.toMap
         val mPutResult: Map[K, Future[Unit]] = store.multiPut(pairMap.mapValues(_._2))
         mPutResult.map { case (k, funit) =>
           (k, funit.map { _ => pairMap(k)._1 })
         }
       }
 
+    val missingFn: K => Future[Option[V]] = { k =>
+      getFailures.flatMap { failures =>
+        Future.exception(failures.getOrElse(k, new MissingValueException[K](k)))
+      }
+    }
+
     /**
-     * Combine original keys with result after put.
+     * Combine original keys with result after put and errors with get.
      */
-    FutureOps.liftFutureValues(keySet, mPutResultFut, getFailures)
+    FutureOps.liftFutureValues(keySet, mPutResultFut, missingFn)
   }
 
   /**
    * Collects keyed futures, partitioning out the failures.
    */
-  private[algebra] def collectWithFailures[K, V](fs:Map[K, Future[V]]): (Future[Seq[(K, V)]], Map[K, Future[Nothing]]) = {
+  private[algebra] def collectWithFailures[K, V](fs: Map[K, Future[V]]): (Future[Array[(K, V)]], Future[Map[K, Throwable]]) = {
     if (fs.isEmpty) {
-      (Future.value(Seq.empty[(K, V)]), Map.empty[K, Future[Nothing]])
+      (Future.value(Array.empty[(K, V)]), Future.value(Map.empty[K, Throwable]))
     } else {
       val fsSize = fs.size
-      val results = new AtomicReferenceArray[Either[(K,Future[Nothing]), (K, V)]](fsSize)
+      val results = new AtomicReferenceArray[(K, Either[Throwable, V])](fsSize)
       val countdown = new AtomicInteger(fsSize)
       val successCount = new AtomicInteger(0)
-      val pSuccess = new Promise[Seq[(K, V)]]
-      var failures = Map.empty[K, Future[Nothing]]
+      val pSuccess = new Promise[Array[(K, V)]]
+      val pFailures = new Promise[Map[K, Throwable]]
 
       def collectResults() = {
         if (countdown.decrementAndGet() == 0) {
           val successArray = new Array[(K, V)](successCount.get())
           var si = 0
           var ri = 0
+          var failures = Map.empty[K, Throwable]
           while (ri < fsSize) {
             results.get(ri) match {
-              case Right(kv) =>
-                successArray(si) = kv
+              case (k, Right(v)) =>
+                successArray(si) = k -> v
                 si += 1
-              case Left(kv) =>
-                failures = failures + kv
+              case (k, Left(t)) =>
+                failures = failures + (k -> t)
             }
             ri += 1
           }
           pSuccess.setValue(successArray)
+          pFailures.setValue(failures)
         }
       }
 
       for (((k, f), i) <- fs.iterator.zipWithIndex) {
```

Comment:
Implemented this, it does add a bit of complexity. Also does avoid one wrapping/unwraping for each item in the map.